File: problems/chapter06/mshibatatt/004.py
# http://poj.org/problem?id=2456

def is_feasible(distance, N, C, x):
    last_position = 0
    cow_number = 1
    for i in range(1, N):
        if x[i]-x[last_position] < distance:
            last_position = i
            cow_number += 1
            if cow_number == C:
                return True
    return False

def main():
    N, C = map(int, input().split())
    x = [int(input()) for _ in range(N)]
    x.sort()

    left = 0
    right = max(x)
    while right - left > 1:
        mid = left + (right - left)//2
        if is_feasible(mid, N, C, x):
            left = mid
        else:
            right = mid

    print(left)

if __name__=='__main__':
    main()

File: problems/chapter06/miiitomi/004.cpp
// POJ #2456 Aggressive cows
// http://poj.org/problem?id=2456

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N, C;
vector<int> x;

bool f(int m) {
    int last = 0;
    int counter = 1;
    for (int i = 1; i < N; i++) {
        if (x[i] - x[last] >= m) {
            last = i;
            counter += 1;
        }
        if (counter == C) {
            return true;
        }
    }
    return false;
}

int main() {
    cin >> N >> C;
    x.resize(N);
    for (int i = 0; i < N; i++) cin >> x[i];
    
    sort(x.begin(), x.end());

    int left = 1;
    int right = x[N-1] + 1;

    while (right - left > 1) {
        int mid = (left + right) / 2;
        if (f(mid)) {
            left = mid;
        } else {
            right = mid;
        }
    }

    cout << left << endl;
}
